Drop Custom toJS() / toJSNewlyCreated() functions for Document, XMLDocument & HTMLDocument

cdumez opened the pull request:
#### 00591b21d6b64f611ab3c4c9988074edd9f03089
<pre>
Drop Custom toJS() / toJSNewlyCreated() functions for Document, XMLDocument &amp; HTMLDocument
<a href="https://bugs.webkit.org/show_bug.cgi?id=300427">https://bugs.webkit.org/show_bug.cgi?id=300427</a>

Reviewed by Ryosuke Niwa.

Drop Custom toJS() / toJSNewlyCreated() functions for Document, XMLDocument &amp; HTMLDocument
by extending the bindings generator support for these.

I had to add `[SkipVTableValidation]` to XMLDocument.idl and HTMLDocument.idl because the
generated vtable check would fail, presumably due to them having subclasses. Note that the
original custom bindings code had no vtable validation check so this check would have been
new after my change anyway.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDocumentCustom.cpp:
(WebCore::createNewDocumentWrapper): Deleted.
(WebCore::toJSNewlyCreated): Deleted.
(WebCore::toJS): Deleted.
* Source/WebCore/bindings/js/JSHTMLDocumentCustom.cpp: Removed.
* Source/WebCore/bindings/js/JSXMLDocumentCustom.cpp: Removed.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/XMLDocument.idl:
* Source/WebCore/html/HTMLDocument.idl:

Canonical link: <a href="https://commits.webkit.org/301291@main">https://commits.webkit.org/301291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c75a8bdf2cb206b38b14fdcdd09581201ecaf23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125374 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35780 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132224 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77249 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c21f5c9c-2cfe-4d7a-ab14-0aa372239de2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127251 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53604 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95458 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/63390 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c0149d85-6ea6-4afb-afbb-102c64b08327) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112122 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75997 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c9241586-3866-4f71-ac80-38322a185ca0) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35426 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30291 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75701 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106296 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30515 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134909 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52176 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39954 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103931 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108342 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103690 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26430 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Ryosuke Niwa; Running compile-webkit") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49058 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27353 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49299 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19657 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52072 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57851 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51428 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54784 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53122 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5555,7 +5555,12 @@ END
-        push (@implContent, "    UNUSED_PARAM(lexicalGlobalObject);\n");
+        if ($implType =~ /Document$/) {
+            AddToImplIncludes("JSDocumentCustom.h");
+            push(@implContent, "    reportMemoryForDocumentIfFrameless(*lexicalGlobalObject, impl.get());\n");
```
Hm... it's unfortunate that we need to hard-code this here.
I wonder if we can make it a more generic mechanism somehow?

cdumez committed:
Drop Custom toJS() / toJSNewlyCreated() functions for Document, XMLDocument & HTMLDocument
https://bugs.webkit.org/show_bug.cgi?id=300427

Reviewed by Ryosuke Niwa.

Drop Custom toJS() / toJSNewlyCreated() functions for Document, XMLDocument & HTMLDocument
by extending the bindings generator support for these.

I had to add `[SkipVTableValidation]` to XMLDocument.idl and HTMLDocument.idl because the
generated vtable check would fail, presumably due to them having subclasses. Note that the
original custom bindings code had no vtable validation check so this check would have been
new after my change anyway.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDocumentCustom.cpp:
(WebCore::createNewDocumentWrapper): Deleted.
(WebCore::toJSNewlyCreated): Deleted.
(WebCore::toJS): Deleted.
* Source/WebCore/bindings/js/JSHTMLDocumentCustom.cpp: Removed.
* Source/WebCore/bindings/js/JSXMLDocumentCustom.cpp: Removed.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/XMLDocument.idl:
* Source/WebCore/html/HTMLDocument.idl:

Canonical link: https://commits.webkit.org/301291@main
